lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp

```diff
@@ -1647,35 +1652,133 @@ namespace Js
+                JavascriptString *partType = JavascriptString::NewCopySz(partKindStr, scriptContext);
+
+                DynamicObject* part = scriptContext->GetLibrary()->CreateObject();
+                JavascriptOperators::InitProperty(part, PropertyIds::type, partType);
+                JavascriptOperators::InitProperty(part, PropertyIds::value, partValue);
+
+                ret->SetItem(i, part, PropertyOperationFlags::PropertyOperation_None);
+                i += 1;
+            }
+
+            return ret;
+        }
+        else
+        {
+            FormatDateTime(dtf, x, formatted, required);
+            return JavascriptString::NewWithBuffer(formatted, required - 1, scriptContext);
+        }
 #endif
-        return scriptContext->GetLibrary()->GetNull();
+    }
+
+    Var IntlEngineInterfaceExtensionObject::EntryIntl_GetPatternForSkeleton(RecyclableObject *function, CallInfo callInfo, ...)
+    {
+#ifdef INTL_ICU
+        EngineInterfaceObject_CommonFunctionProlog(function, callInfo);
+        INTL_CHECK_ARGS(args.Info.Count == 3 && JavascriptString::Is(args.Values[1]) && JavascriptString::Is(args.Values[2]));
+
+        JavascriptString *locale = JavascriptString::UnsafeFromVar(args.Values[1]);
+        JavascriptString *skeleton = JavascriptString::UnsafeFromVar(args.Values[2]);
+        utf8::WideToNarrow locale8(locale->GetSz(), locale->GetLength());
+
+        int patternLen = GetPatternForSkeleton(locale8, skeleton->GetSz());
+        char16 *pattern = RecyclerNewArrayLeaf(scriptContext->GetRecycler(), char16, patternLen);
+        GetPatternForSkeleton(locale8, skeleton->GetSz(), pattern, patternLen);
+
+        return JavascriptString::NewWithBuffer(pattern, patternLen - 1, scriptContext);
+#else
+        AssertOrFailFastMsg(false, "GetPatternForSkeleton is not implemented outside of ICU");
+        return nullptr;
 #endif
     }
 
-    /*
-    *   This function validates the timeZone passed by user has defined in IsValidTimeZoneName() section
-    *   of ECMA-402 dated June 2015.
-    *   Returns true if timeZoneId is a valid zone or link name of the IANA time zone database
-    */
+    // given a timezone name as an argument, this will return a canonicalized version of that name, or undefined if the timezone is invalid
     Var IntlEngineInterfaceExtensionObject::EntryIntl_ValidateAndCanonicalizeTimeZone(RecyclableObject* function, CallInfo callInfo, ...)
     {
         EngineInterfaceObject_CommonFunctionProlog(function, callInfo);
+        INTL_CHECK_ARGS(args.Info.Count == 2 && JavascriptString::Is(args.Values[1]));
 
-        // Return false if timeZoneId is not string
-        if (args.Info.Count < 2 || !JavascriptString::Is(args.Values[1]))
-        {
-            AssertMsg(false, "Need valid timeZoneId");
-            return scriptContext->GetLibrary()->GetUndefined();
-        }
+        JavascriptString *tz = JavascriptString::FromVar(args.Values[1]);
```

Comment:
Should this be UnsafeFromVar to match with the style of previous functions? (Personally, I like the TryFromVar and null-check pattern better than Is followed by UnsafeFromVar, but at least we should be consistent.)

Reply:
I think that I didn't do this because this variable is used by WinGlob code too, and I wanted to touch as little WG logic as possible? I can probably change it.